contents/functionalProgramming/intro.md

```diff
@@ -0,0 +1,132 @@
+
+- Haskell
+- Clojure
+- Elixir
+- Elm
+- Scala
+- ML (Meta Language) family of languages
+
+While functional programming can be implemented by languages like Java, the languages listed above encourage the functional programming paradigm such as [pure functions](#pure-functions) or even enforce them in the case of Haskell.
+
+## Concepts in Functional Programming
+
+#### Pure Functions
+
+Pure functions is one of the concepts in functional programming. To call a function `pure`, it needs to satisfy two conditions:
+
+1. [Idempotence](https://en.wikipedia.org/wiki/Idempotence) - The function should return the same output for the same input when invoked arbitrarily many times.
+1. No side effects - The function should not cause side effects like mutating global state.
+
+To illustrate an example of a pure function. Consider this simple example in Python:
+
+```python
+# Pure Function
+def add(x, y):
+    return x + y
+```
+
+Contrast this with an impure function:
+
+```python
+# Mutate state
+y=0
+def addXToY(x):
+    global y
+    y = y + x
+    return y
+```
+We see that the impure function violates both conditions. Running `addXToY` multiple times with the same input X would give different value.
+
+#### Immutability
+
+Another concept in functional programming is immutability. Immutability prevents an object's state to be change after it is created. When we assign `x=1`, it does not make mathematical sense to have this expression `x:=x+1`.
+
+Having immutability built into functional programming languages like Haskell also helps to ensure that functions are `pure` because mutable variables and states can introduce side-effects.
+
+To know more about immutability in functional languages, you can take a look at:
+
+- [Immutability in Haskell](https://mmhaskell.com/blog/2017/1/9/immutability-is-awesome)
+- [Immutability in Elm](http://elmprogramming.com/immutability.html)
+- [Why does immutability enable functional programming](https://stackoverflow.com/questions/12207757/why-do-immutable-objects-enable-functional-programming)
+
+## Techniques in Functional Programming
+
+#### Recursion
+
+Pure functional languages do not have loop constructs like imperative languages. This is because `Iteration` usually involves state mutation per iteration. Since functional programming avoids state changes, `Recursion` is a commonly used [technique](https://www.quora.com/Why-dont-pure-functional-programming-languages-provide-a-loop-construct) to replace `Iteration`. An example of replacing state-mutating iterative code to a pure functional recusive code is shown below:
+
+```Python
+# Iterative
+def getSumOfListIter(listOfNumbers):
+    sum = 0
+    for number in listOfNumbers:
+        sum += number
+    return sum
+
+# Recursive
+def getSumOfListRec(listOfNumbers):
+    if len(listOfNumbers) == 0:
+        return 0
+    return listOfNumbers[0] + getSumOfListRec(listOfNumbers[1:])
+```
```

Comment:
Give as two code segments. Otherwise this code will be rejected by Python?

Reply:
@damithc, can you clarify this?

Reply:
I ran the code and it works. The functions are named differently as well.